implement performance logging

rkusa opened the pull request:
First attempt at implementing #27.

See comments for further details.

Example logs:

```
2021-10-12T14:35:57.238404900+02:00 INFO dcs_grpc_server::stats - avg. TPS = 0.93 | max. TPS = 15.05 | avg. wait time = 14ms | total wait time = 820ms | pending requests = 0 | max. pending requests = 0
2021-10-12T14:36:57.239474800+02:00 INFO dcs_grpc_server::stats - avg. TPS = 0.70 | max. TPS = 15.05 | avg. wait time = 14ms | total wait time = 1.23s | pending requests = 0 | max. pending requests = 0
2021-10-12T14:37:57.240512600+02:00 INFO dcs_grpc_server::stats - avg. TPS = 0.62 | max. TPS = 15.05 | avg. wait time = 14ms | total wait time = 1.67s | pending requests = 0 | max. pending requests = 0
```

rurounijones commented on src/stats.rs:
```diff
@@ -0,0 +1,160 @@
+    summary: Arc<Mutex<Summary>>,
+}
+
+struct Summary {
```
All of these values are for the last minute right?

rkusa replied:
The values are the last second.
The summary is calculated every second, but it is only logged every minute. I am calculating it every second to keep track of e.g. the highest tps (as it would be the highest transactions per minute otherwise). Does this make sense?

rkusa commented on src/stats.rs:
```diff
@@ -0,0 +1,160 @@
+    /// Average time spent waiting for MSE calls to complete.
+    wait_time_average: Duration,
+    /// Total time spent waiting for MSE calls to complete.
+    wait_time_total: Duration,
```
True that, I'll think about a possible update tomorrow 👍

rkusa replied:
I've update the PR to track the blocking time instead (in 2fdfeb52160d754fd49b0bcfc1a8c83448ca66cf).

Example:

```
2021-10-13T22:03:15.723161600+02:00 INFO dcs_grpc_server::stats - avg. TPS = 1.35 | max. TPS = 7.97 | avg. blocking time = 229.955µs | total blocking time = 6.2088ms | pending requests = 0 | max. pending requests = 0
2021-10-13T22:03:36.714055300+02:00 INFO dcs_grpc_server::stats - avg. TPS = 1.07 | max. TPS = 7.97 | avg. blocking time = 214.986µs | total blocking time = 9.4594ms | pending requests = 0 | max. pending requests = 0
```

rurounijones replied:
I am not sure I understand the difference? I think the valuable value is "The average blocking time of all requests processed in the last minute". However you name that.

rurounijones commented:
I ran a mission with 300 odd units moving and fighting with the standard TPS settings and recorded the following (Note, I cleaned out some fields to aid readability)

I was watching on Jupiter (Which streams the units *and* the events) and I didn't notice any slowdowns. The units seemed to be updating at a steady rate.

```
23:02:22 avg. TPS = 148.67 | max. TPS = 602.25 | avg. blocking time = 159.406µs | total blocking time = 1.4223901s  | pending requests = 1 | max. pending requests = 1
23:03:23 avg. TPS = 241.67 | max. TPS = 602.25 | avg. blocking time = 150.863µs | total blocking time = 4.4121729s  | pending requests = 0 | max. pending requests = 1
23:04:23 avg. TPS = 273.90 | max. TPS = 602.25 | avg. blocking time = 145.981µs | total blocking time = 7.2379263s  | pending requests = 0 | max. pending requests = 1
23:05:24 avg. TPS = 288.77 | max. TPS = 602.25 | avg. blocking time = 143.845µs | total blocking time = 10.0526405s | pending requests = 0 | max. pending requests = 1
23:06:24 avg. TPS = 295.19 | max. TPS = 602.25 | avg. blocking time = 142.29µs  | total blocking time = 12.6855659s | pending requests = 0 | max. pending requests = 1
23:07:25 avg. TPS = 297.11 | max. TPS = 602.25 | avg. blocking time = 141.089µs | total blocking time = 15.2178043s | pending requests = 0 | max. pending requests = 1
23:08:26 avg. TPS = 294.58 | max. TPS = 602.25 | avg. blocking time = 141.033µs | total blocking time = 17.6156224s | pending requests = 1 | max. pending requests = 1
23:09:26 avg. TPS = 286.90 | max. TPS = 602.25 | avg. blocking time = 140.619µs | total blocking time = 19.5275539s | pending requests = 0 | max. pending requests = 1
23:10:27 avg. TPS = 280.23 | max. TPS = 602.25 | avg. blocking time = 141.088µs | total blocking time = 21.5484418s | pending requests = 0 | max. pending requests = 1
23:11:27 avg. TPS = 273.02 | max. TPS = 602.25 | avg. blocking time = 142.042µs | total blocking time = 23.4631618s | pending requests = 0 | max. pending requests = 1
23:12:27 avg. TPS = 265.44 | max. TPS = 602.25 | avg. blocking time = 141.7µs   | total blocking time = 25.0138093s | pending requests = 0 | max. pending requests = 1
23:13:28 avg. TPS = 257.92 | max. TPS = 602.25 | avg. blocking time = 141.466µs | total blocking time = 26.4903517s | pending requests = 0 | max. pending requests = 1
23:14:29 avg. TPS = 249.23 | max. TPS = 602.25 | avg. blocking time = 141.048µs | total blocking time = 27.6663441s | pending requests = 0 | max. pending requests = 1
```

The total blocking time appears to be climbing constantly and seemed to be a sum of the whole mission. I was expecting this value to be the total blocking time for the last minute only. Having it for the entire mission duration is not very useful.

rurounijones commented on src/stats.rs:
```diff
@@ -0,0 +1,160 @@
+                    format_duration(summary.wait_time_total),
+                    summary.pending_current,
+                    summary.pending_highest,
+                )
```
I think what is there is fine. I don't have any suggestions that would make it much more readable.

rurounijones commented on src/stats.rs:
```diff
@@ -0,0 +1,160 @@
+            if last_logged.elapsed() > log_interval {
+                last_logged = Instant::now();
+
+                log::info!(
```
Having run the tests can we get the performance logs to also be put in the `dcs.log` ? I didn't see them during my testing. It will help server admins having it inline with everything else.

rurounijones commented:
Ran another one with througput limit set to 50.

```
09:51:56 INFO avg. TPS = 1.48 | max. TPS = 4.02 | avg. blocking time = 625.988µs | total blocking time = 55.713ms | pending requests = 0 | max. pending requests = 0
09:52:57 INFO avg. TPS = 18.55 | max. TPS = 63.79 | avg. blocking time = 158.002µs | total blocking time = 354.7167ms | pending requests = 0 | max. pending requests = 1
09:53:58 INFO avg. TPS = 28.36 | max. TPS = 64.00 | avg. blocking time = 147.343µs | total blocking time = 760.5896ms | pending requests = 0 | max. pending requests = 1
09:54:58 INFO avg. TPS = 33.01 | max. TPS = 64.35 | avg. blocking time = 143.562µs | total blocking time = 1.1469244s | pending requests = 0 | max. pending requests = 1
09:55:58 INFO avg. TPS = 35.86 | max. TPS = 64.71 | avg. blocking time = 141.772µs | total blocking time = 1.5354001s | pending requests = 0 | max. pending requests = 1
09:56:59 INFO avg. TPS = 37.87 | max. TPS = 64.71 | avg. blocking time = 141.661µs | total blocking time = 1.9474263s | pending requests = 0 | max. pending requests = 1
09:58:00 INFO avg. TPS = 39.36 | max. TPS = 64.71 | avg. blocking time = 140.762µs | total blocking time = 2.3491974s | pending requests = 0 | max. pending requests = 1
09:59:00 INFO avg. TPS = 40.52 | max. TPS = 64.78 | avg. blocking time = 140.486µs | total blocking time = 2.7555265s | pending requests = 0 | max. pending requests = 1
10:00:01 INFO avg. TPS = 41.32 | max. TPS = 64.78 | avg. blocking time = 139.897µs | total blocking time = 3.150622s | pending requests = 0 | max. pending requests = 1
10:01:01 INFO avg. TPS = 41.95 | max. TPS = 64.78 | avg. blocking time = 140.161µs | total blocking time = 3.5570321s | pending requests = 0 | max. pending requests = 1
10:02:02 INFO avg. TPS = 42.47 | max. TPS = 64.78 | avg. blocking time = 140.722µs | total blocking time = 3.9801951s | pending requests = 0 | max. pending requests = 1
10:03:03 INFO avg. TPS = 42.91 | max. TPS = 64.78 | avg. blocking time = 140.676µs | total blocking time = 4.3890999s | pending requests = 0 | max. pending requests = 1
10:04:04 INFO avg. TPS = 43.20 | max. TPS = 64.78 | avg. blocking time = 140.895µs | total blocking time = 4.7963813s | pending requests = 0 | max. pending requests = 1
10:05:04 INFO avg. TPS = 43.51 | max. TPS = 64.78 | avg. blocking time = 141.22µs | total blocking time = 5.2109056s | pending requests = 0 | max. pending requests = 1
10:06:04 INFO avg. TPS = 43.71 | max. TPS = 64.78 | avg. blocking time = 141.253µs | total blocking time = 5.6063605s | pending requests = 0 | max. pending requests = 1
```

Definitely slower updates on Jupiter but I don't see any requests piling up. Does the "Requests" metric not include anything in the streaming API?

rkusa commented:
Changes:

- I've updated the PR according to https://github.com/DCS-gRPC/rust-server/pull/39#discussion_r731064026
- Fixed the stats about the queue size (was indeed incorrectly tracked)
- Parallelized the retrieval of all groups when a units stream is started

I've tested it on a mission with exactly 400 planes. Here are some of the logs as an example:

`throughputLimit = 1` (max. 1 calls every 1s)

```
11:59:30.5 INFO dcs_grpc_server::stats - Calls per second: average=0.93, highest=1.01
11:59:30.5 INFO dcs_grpc_server::stats - Blocking time: average=314.975µs/s, total=18.8994ms
11:59:30.5 INFO dcs_grpc_server::stats - Queue size: average=66.25, biggest=100

12:00:30.5 INFO dcs_grpc_server::stats - Calls per second: average=1.00, highest=1.01
12:00:30.5 INFO dcs_grpc_server::stats - Blocking time: average=271.455µs/s, total=16.2885ms
12:00:30.5 INFO dcs_grpc_server::stats - Queue size: average=111.34, biggest=400

12:01:31.5 INFO dcs_grpc_server::stats - Calls per second: average=1.00, highest=1.01
12:01:31.5 INFO dcs_grpc_server::stats - Blocking time: average=94.902µs/s, total=5.7895ms
12:01:31.5 INFO dcs_grpc_server::stats - Queue size: average=399.97, biggest=400

12:02:32.5 INFO dcs_grpc_server::stats - Calls per second: average=1.00, highest=1.01
12:02:32.5 INFO dcs_grpc_server::stats - Blocking time: average=93.954µs/s, total=5.7316ms
12:02:32.5 INFO dcs_grpc_server::stats - Queue size: average=399.98, biggest=400

12:03:33.5 INFO dcs_grpc_server::stats - Calls per second: average=0.98, highest=1.01
12:03:33.5 INFO dcs_grpc_server::stats - Blocking time: average=91.111µs/s, total=5.5573ms
12:03:33.5 INFO dcs_grpc_server::stats - Queue size: average=400.04, biggest=400
```

The queue size seems to be legit now. In the first interval, `100` corresponds to the 100 groups in my mission (for which we are fetching the units first). In the next interval, the `400` corresponds to the 400 units in the mission. We also see that waiting for all unit updates before updating them again works, because the queue size is never above 400.

`throughputLimit = 600` (max. 18 calls every 0.03s; default)

```
12:05:51.1 INFO dcs_grpc_server::stats - Calls per second: average=149.55, highest=598.85
12:05:51.1 INFO dcs_grpc_server::stats - Blocking time: average=5.749981ms/s, total=345.0694ms
12:05:51.1 INFO dcs_grpc_server::stats - Queue size: average=106.38, biggest=400

12:06:52.1 INFO dcs_grpc_server::stats - Calls per second: average=165.56, highest=590.41
12:06:52.1 INFO dcs_grpc_server::stats - Blocking time: average=5.9585ms/s, total=363.424ms
12:06:52.1 INFO dcs_grpc_server::stats - Queue size: average=119.75, biggest=400

12:07:53.1 INFO dcs_grpc_server::stats - Calls per second: average=161.05, highest=587.43
12:07:53.1 INFO dcs_grpc_server::stats - Blocking time: average=5.767382ms/s, total=351.8704ms
12:07:53.1 INFO dcs_grpc_server::stats - Queue size: average=121.55, biggest=400
```

rurounijones commented on src/stats.rs:
```diff
@@ -0,0 +1,176 @@
+                    "Queue size: average={:.2}, biggest={:.2}",
+                    queue_size_average,
+                    interval_stats.queue_size_highest
+                );
```
I think having it be multi-line and more verbose is the better format.

rurounijones commented:
Looking good. One last nitpick (Note I am being a user here, I am not reading the code).

> 12:01:31.5 INFO dcs_grpc_server::stats - Queue size: average=399.97, biggest=400

Not sure what the .5 means in this case. Is it:
* minute 31 and 5 seconds? 
* minute 31 point 5 (so 51 and 30 seconds)
* minute 31 and 50 seconds?

I don't think we actually need anything more granular than the minute level but if we do log at the second level then I would go for a more standard .00 (%2d) format.
